File: frontend/src/model/ExtendedPresentationElement.ts
import PresentationElement from "./PresentationElement";
import { decodeExtendedSectionPresentationElement } from "./ExtendedSectionPresentationElement";
import { decodeExtendedRuleInvocationPresentationElement } from "./ExtendedRuleInvocationPresentationElement";

export const decodeExtendedPresentationElement = <T extends PresentationElement>(json: any): ExtendedPresentationElement => {
  switch (json.type) {
    case "Section":
      return decodeExtendedSectionPresentationElement(json);
    case "RuleInvocation":
      return decodeExtendedRuleInvocationPresentationElement(json);
    default:
      throw Error("ExtendedPresentationElement type " + json.type + " is not supported");
  }
};

export interface ExtendedPresentationElement {
  presentationElement: PresentationElement;
}

File: compiler/src/main/kotlin/com/piotrwalkusz/smartlaw/compiler/converter/extend/ExtendedPresentationElement.kt
package com.piotrwalkusz.smartlaw.compiler.converter.extend

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.piotrwalkusz.smartlaw.compiler.converter.naturallanguage.model.NaturalLanguageDocumentObject
import com.piotrwalkusz.smartlaw.core.model.presentation.PresentationElement

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes(
        JsonSubTypes.Type(name = "Section", value = ExtendedSectionPresentationElement::class),
        JsonSubTypes.Type(name = "RuleInvocation", value = ExtendedRuleInvocationPresentationElement::class)
)
interface ExtendedPresentationElement<T : PresentationElement, R: NaturalLanguageDocumentObject> {
    val presentationElement: T
    val naturalLanguageDocumentObject: R
}